MCOL-5205 Fix bug from union type in UNION processing

JigaoLuo opened the pull request:
The Jira issue number for this PR is: [MCOL-5205](https://jira.mariadb.org/browse/MCOL-5205)

*NOTE: This project is for the Google Summer of Code 2022.*

# Task

- [x] Pure Int UNION cases
- [x] Tests on pure Int UNION cases
- [x] Tests on UNION and UNION ALL

tntnatbry commented on datatypes/mcs_datatype.h:
```diff
@@ -506,6 +506,74 @@ inline bool isSignedInteger(const datatypes::SystemCatalog::ColDataType type)
   }
 }
 
+inline bool isUnsignedInteger(const datatypes::SystemCatalog::ColDataType type)
```
To avoid code duplication, can you please use `isUnsignedInteger()` call in the above `isUnsigned()` function.

tntnatbry commented on datatypes/mcs_datatype.h:
```diff
@@ -506,6 +506,74 @@ inline bool isSignedInteger(const datatypes::SystemCatalog::ColDataType type)
+    case datatypes::SystemCatalog::UBIGINT:
+    {
+      type = datatypes::SystemCatalog::DECIMAL;
+      colWidth = 16;
```
Here, the precision should also be set, to `datatypes::INT128MAXPRECISION`, and the scale to 0.

tntnatbry commented on utils/dataconvert/dataconvert.cpp:
```diff
@@ -2984,25 +2984,28 @@ void DataConvert::joinColTypeForUnion(datatypes::SystemCatalog::TypeHolderStd& u
-            unionedType.colDataType = type.colDataType;
-            unionedType.colWidth = type.colWidth;
+            // No handling. Fallthrough.
           }
```
When `colDataType` is same, then the condition `type.colWidth > unionedType.colWidth` should be checked and should appropriately set the width.

tntnatbry commented on utils/dataconvert/dataconvert.cpp:
```diff
@@ -2984,25 +2984,28 @@ void DataConvert::joinColTypeForUnion(datatypes::SystemCatalog::TypeHolderStd& u
+          if (isDecimal(type.colDataType))
           {
             unionedType.colDataType = datatypes::SystemCatalog::DECIMAL;
           }
```
This does not handle all edge cases related to DECIMAL. For example the case when `type=DECIMAL` with width 2 bytes and `unionedType` was a BIGINT, this will incorrectly set the resulting `unionedType` to DECIMAL. Signedness and width should be appropriately considered for the case when `type` and/or `unionedType` are DECIMAL.

tntnatbry replied:
Can you please address the above comment?

tntnatbry commented on mysql-test/columnstore/basic/t/union.test:
```diff
@@ -17,46 +17,279 @@ CREATE TABLE t1 (a INT, b INT UNSIGNED);
 SELECT * FROM (SELECT b FROM t1 UNION SELECT a FROM t1) tu ORDER BY b;
+INSERT INTO t1 VALUES (-1,1);
+SELECT * FROM (SELECT a FROM t1 UNION SELECT b FROM t1) tu ORDER BY a;
+SELECT * FROM (SELECT b FROM t1 UNION SELECT a FROM t1) tu ORDER BY b;
```
What is the point of the above 3 additions? They are repetitions of the previous 3 statements.

JigaoLuo replied:
I want to test the duplicate elimination of `UNION` by inserting (1,1) again. The output should be the same in table t1. This test is helpful to check the buffering status in UNION processing.

tntnatbry commented on mysql-test/columnstore/basic/t/union.test:
```diff
@@ -17,46 +17,279 @@ CREATE TABLE t1 (a INT, b INT UNSIGNED);
+INSERT INTO t1 VALUES (-2,2);
+SELECT * FROM (SELECT * FROM t1 UNION SELECT * FROM t1) tu;
+SELECT * FROM (SELECT a FROM t1 UNION SELECT b FROM t1) tu ORDER BY a;
+SELECT * FROM (SELECT b FROM t1 UNION SELECT a FROM t1) tu ORDER BY b;
```
Just INSERT (-2, 2) at the start when the table is created, and remove the duplicated SELECTs.

tntnatbry commented on mysql-test/columnstore/basic/t/union.test:
```diff
@@ -17,46 +17,279 @@ CREATE TABLE t1 (a INT, b INT UNSIGNED);
+SELECT * FROM (SELECT a FROM t1 UNION SELECT b FROM t1) tu ORDER BY a;
+SELECT * FROM (SELECT b FROM t1 UNION SELECT a FROM t1) tu ORDER BY b;
+INSERT INTO t1 VALUES (-2,2);
+SELECT * FROM (SELECT * FROM t1 UNION SELECT * FROM t1) tu;
```
There should be an ORDER BY clause to make the result set deterministic.

tntnatbry commented on mysql-test/columnstore/basic/t/union.test:
```diff
@@ -17,46 +17,279 @@ CREATE TABLE t1 (a INT, b INT UNSIGNED);
+INSERT INTO t1 VALUES (1, 1), (1, 1), (2, 1), (2, 1), (3, 2), (4, 2), (5, 2);
+SELECT * FROM (SELECT a FROM t1 UNION SELECT b FROM t1) tu ORDER BY a;
+SELECT * FROM (SELECT b FROM t1 UNION SELECT a FROM t1) tu ORDER BY b;
+SELECT * FROM (SELECT * FROM t1 UNION SELECT * FROM t1) tu;
```
There should be an ORDER BY clause to make the result set deterministic.

tntnatbry commented on mysql-test/columnstore/basic/t/union.test:
```diff
@@ -17,46 +17,279 @@ CREATE TABLE t1 (a INT, b INT UNSIGNED);
+INSERT INTO t2 VALUES (CAST(0x8000000000000000 AS SIGNED)+2),(-1),(0),(1),(0x7FFFFFFFFFFFFFFF);
+SELECT * FROM (SELECT a FROM t1 UNION ALL SELECT a FROM t2) tu ORDER BY a;
+SELECT * FROM (SELECT a FROM t2 UNION ALL SELECT a FROM t1) tu ORDER BY a;
+DROP TABLE t1,t2;
```
In the above 4 cases, the same t1 is dropped and inserted multiple times, you can create t1 just once and reuse.

tntnatbry commented on mysql-test/columnstore/basic/t/union.test:
```diff
@@ -17,46 +17,279 @@ CREATE TABLE t1 (a INT, b INT UNSIGNED);
+INSERT INTO t2 VALUES (CAST(0x8000000000000000 AS SIGNED)+2),(-1),(0),(1),(0x7FFFFFFFFFFFFFFF);
+SELECT * FROM (SELECT a FROM t1 UNION ALL SELECT a FROM t2) tu ORDER BY a;
+SELECT * FROM (SELECT a FROM t2 UNION ALL SELECT a FROM t1) tu ORDER BY a;
+DROP TABLE t1,t2;
```
Again, in the above 3 cases, the same t1 is dropped and inserted multiple times, you can create t1 just once and reuse.

tntnatbry commented on mysql-test/columnstore/basic/t/union.test:
```diff
@@ -17,46 +17,279 @@ CREATE TABLE t1 (a INT, b INT UNSIGNED);
+INSERT INTO t2 VALUES (CAST(0x8000000000000000 AS SIGNED)+2),(-1),(0),(1),(0x7FFFFFFFFFFFFFFF);
+SELECT * FROM (SELECT a FROM t1 UNION ALL SELECT a FROM t2) tu ORDER BY a;
+SELECT * FROM (SELECT a FROM t2 UNION ALL SELECT a FROM t1) tu ORDER BY a;
+DROP TABLE t1,t2;
```
Reuse t1 in the above 2 cases.

tntnatbry commented on mysql-test/columnstore/basic/t/union.test:
```diff
@@ -17,46 +17,279 @@ CREATE TABLE t1 (a INT, b INT UNSIGNED);
+SELECT * FROM (SELECT a FROM t1 UNION ALL SELECT a FROM t2) tu ORDER BY a;
+SELECT * FROM (SELECT a FROM t2 UNION ALL SELECT a FROM t1) tu ORDER BY a;
+DROP TABLE t1,t2;
+
```
Please reuse t1 in the above 4 scenarios.

tntnatbry commented on mysql-test/columnstore/basic/t/union.test:
```diff
@@ -17,46 +17,279 @@ CREATE TABLE t1 (a INT, b INT UNSIGNED);
+INSERT INTO t2 VALUES (0),(1),(0xFFFFFFFFFFFFFFFF-2);
+SELECT * FROM (SELECT a FROM t1 UNION ALL SELECT a FROM t2) tu ORDER BY a;
+SELECT * FROM (SELECT a FROM t2 UNION ALL SELECT a FROM t1) tu ORDER BY a;
+DROP TABLE t1,t2;
```
Please reuse t1 in the above 3 scenarios.

tntnatbry commented on mysql-test/columnstore/basic/t/union.test:
```diff
@@ -17,46 +17,279 @@ CREATE TABLE t1 (a INT, b INT UNSIGNED);
+INSERT INTO t2 VALUES (0),(1),(0xFFFFFFFFFFFFFFFF-2);
+SELECT * FROM (SELECT a FROM t1 UNION ALL SELECT a FROM t2) tu ORDER BY a;
+SELECT * FROM (SELECT a FROM t2 UNION ALL SELECT a FROM t1) tu ORDER BY a;
+DROP TABLE t1,t2;
```
Please reuse t1 in the above 2 scenarios.

cakebytheoceanLuo commented:
I notice there is a similar case setting the type to DECIMAL: https://github.com/mariadb-corporation/mariadb-columnstore-engine/pull/2526/files#diff-58e39e2454e28831cbe2e09724a4d3754cffe87211028865a93e0ad55df03210R2980-R2985

tntnatbry commented on datatypes/mcs_datatype.h:
```diff
@@ -471,18 +471,29 @@ inline bool isDecimal(const datatypes::SystemCatalog::ColDataType type)
+inline bool isUnsigned(const datatypes::SystemCatalog::ColDataType type)
+{
+  if (isUnsignedInteger(type)) 
+    return true;
```
Please add a line break after this return.

tntnatbry commented on mysql-test/columnstore/basic/t/union.test:
```diff
@@ -15,48 +15,253 @@ USE mcs_union;
 CREATE TABLE t1 (a INT, b INT UNSIGNED);
 INSERT INTO t1 VALUES (-1,1);
+INSERT INTO t1 VALUES (-1,1);
+INSERT INTO t1 VALUES (-2,2);
```
You can combine the 3 INSERTs into a single INSERT: `INSERT INTO t1 VALUES (-1, 1), (-1, 1), (-2, 2);`

tntnatbry commented on mysql-test/columnstore/basic/t/union.test:
```diff
@@ -15,48 +15,253 @@ USE mcs_union;
+CREATE TABLE t2 (a INT UNSIGNED);
+INSERT INTO t2 VALUES (0),(1),(0xFFFFFF-2);
+SELECT * FROM (SELECT a FROM t1 UNION ALL SELECT a FROM t2) tu ORDER BY a;
+SELECT * FROM (SELECT a FROM t2 UNION ALL SELECT a FROM t1) tu ORDER BY a;
```
We should also add DECIMAL test cases:
- Both sides DECIMAL with same signedness and precision.
- Both sides DECIMAL with same signedness and different precision.
- Both sides DECIMAL with different signedness and same precision.
- Both sides DECIMAL with different signedness and different precision.
- One side DECIMAL and other side integer type. Again, there should be variation in test cases in DECIMAL precision, integer types, and signedness.

JigaoLuo committed:
[MCOL-5205] Fix bug from union type in UNION processing.

This patch fixs the reported JIRA issue MCOL 5205, which consists of a wrong union type from two input Int types. The bug results in wrong unioned answers in CS. The fix includes more INT case discussions. Additionaly, this patch provides detailed unit tests for correctness in UNION processing with Int.

Signed-off-by: Jigao Luo <luojigao@outlook.com>
